### PR TITLE
allow list of keys for showing vertices/edges/faces

### PR DESCRIPTION
**Pull request by Licini**
Allow drawing selection of vertices/edges/faces on `MeshObject`
```python
from compas.datastructures import Mesh
from compas.scene import Scene

mesh = Mesh.from_polyhedron(8)

vertices = mesh.face_vertices(0)
edges = mesh.face_halfedges(0)

scene = Scene()

scene.add(mesh, show_vertices=vertices, show_edges=edges)

scene.draw()
```

image

### CODE REVIEW

File: src/compas_rhino/scene/meshobject.py [hunk 1]
```diff
@@ -176,20 +176,33 @@ def draw(self, color=None, vertexcolors=None, facecolors=None, disjoint=False):
...
+            color = Color.coerce(color) or self.color
+            attr = attributes(name=self.mesh.name, color=color, layer=self.layer)  # type: ignore
 
-        geometry.Transform(transformation_to_rhino(self.worldtransformation))
+            geometry = mesh_to_rhino(
+                self.mesh,
+                color=color,
+                vertexcolors=vertexcolors,
+                facecolors=facecolors,
+                disjoint=disjoint,
```
Comment (Licini) [inline src/compas_rhino/scene/meshobject.py]: Should we remove these kwargs from `draw` itself, and only allow to configure them in `scene.add()` ?

File: src/compas_rhino/scene/meshobject.py [hunk 2]
```diff
@@ -176,20 +176,33 @@ def draw(self, color=None, vertexcolors=None, facecolors=None, disjoint=False):
...
+            )
+
+            geometry.Transform(transformation_to_rhino(self.worldtransformation))
+
+            self._guids += [sc.doc.Objects.AddMesh(geometry, attr)]
+
+        elif self.show_faces:
+            self._guids += self.draw_faces(faces=self.show_faces, color=self.facecolor)
```
Comment (Licini) [inline src/compas_rhino/scene/meshobject.py]: Currently there is an bug for this function, I created an issue here #1277

Comment (tomvanmele) [reply src/compas_rhino/scene/meshobject.py]: okay will have a look

**Comment by Licini** [comment]
@tomvanmele What do you think now? See below inline comments:

File: src/compas_rhino/scene/_helpers.py [hunk 3]
```diff
@@ -45,7 +45,7 @@ def ngon(v):
     if v < 3:
         return
     if v == 3:
-        return [0, 1, 2, 2]
+        return [0, 1, 2]
```
Comment (Licini) [inline src/compas_rhino/scene/_helpers.py:48]: fixed draw_face issue here

File: src/compas_rhino/scene/meshobject.py [hunk 4]
```diff
@@ -37,10 +37,25 @@ class MeshObject(RhinoSceneObject, BaseMeshObject):
...
+    group : str, optional
+        The name of the group to add the mesh components. The group will be created if not already present.
+        Default is ``None``.
+
     """
 
     def __init__(self, mesh, **kwargs):
         super(MeshObject, self).__init__(mesh=mesh, **kwargs)
+        self.disjoint = kwargs.get("disjoint", False)
+        self.group = kwargs.get("group", None)
```
Comment (Licini) [inline src/compas_rhino/scene/meshobject.py]: Make these two formal attributes here. (another question is that should they go to core?)

File: src/compas_rhino/scene/meshobject.py [hunk 5]
```diff
@@ -37,10 +37,25 @@ class MeshObject(RhinoSceneObject, BaseMeshObject):
...
+        self._guid_mesh = None
+        self._guids_faces = None
+        self._guids_edges = None
+        self._guids_vertices = None
```
Comment (Licini) [inline src/compas_rhino/scene/meshobject.py:54]: more granular guids to track different things.

File: src/compas_rhino/scene/meshobject.py [hunk 6]
```diff
@@ -149,16 +164,8 @@ def clear_facelabels(self):
...
-
-        Parameters
-        ----------
-        color : tuple[int, int, int], optional
-            The color of the mesh.
-            Default is the value of :attr:`MeshObject.default_color`.
-        disjoint : bool, optional
-            If True, draw the faces of the mesh with disjoint vertices.
+    def draw(self):
+        """Draw the mesh or its components in Rhino.
```
Comment (Licini) [inline src/compas_rhino/scene/meshobject.py:162]: removed here the optional inputs, the rational is that no one should call `obj.draw()` directly, one is supposed to always use `scene.draw()` instead. These options are either passed in at `scene.add()` or edited through the object.

File: src/compas_rhino/scene/meshobject.py [hunk 7]
```diff
@@ -37,10 +37,32 @@ class MeshObject(RhinoSceneObject, BaseMeshObject):
...
+        self._guids_faces = None
+        self._guids_edges = None
+        self._guids_vertices = None
+        self._guids_vertexnormals = None
+        self._guids_facenormals = None
+        self._guids_vertexlabels = None
+        self._guids_edgelabels = None
+        self._guids_facelabels = None
+        self._guids_spheres = None
+        self._guids_pipes = None
```
Comment (Licini) [inline src/compas_rhino/scene/meshobject.py:61]: added guids tracking for all the rest too

File: CHANGELOG.md [hunk 8]
```diff
@@ -9,10 +9,23 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+* Changed `compas.scene.Mesh`'s `show_vertices`, `show_edges`, `show_faces` to optionally accept a sequence of keys.
+* Changed `compas.scene.Graph`'s `show_nodes`, `show_edges` to optionally accept a sequence of keys.
+* Changed `compas.scene.VolMesh`'s `show_vertices`, `show_edges`, `show_faces`, `show_cells` to optionally accept a sequence of keys.
+
 ### Removed
 
+* Removed kwargs from `compas_rhino.scene.MeshObject.draw`.
+* Removed kwargs from `compas_rhino.scene.GraphObject.draw`.
+* Removed kwargs from `compas_rhino.scene.VolMeshObject.draw`.
+
```
Comment (Licini) [inline CHANGELOG.md]: Aligned also GraphObject and VolMeshObject to same principle

Comment (Licini) [reply CHANGELOG.md]: @tomvanmele This PR is ready for review again, could you take another look?

File: src/compas/scene/graphobject.py [hunk 9]
```diff
@@ -30,9 +30,9 @@ class GraphObject(SceneObject):
         The size of the nodes. Default is ``1.0``.
     edgewidth : float
         The width of the edges. Default is ``1.0``.
-    show_nodes : bool
+    show_nodes : Union[bool, sequence[float]]
```
Comment (tomvanmele) [inline src/compas/scene/graphobject.py:33]: would use "|" instead of "Union" in the docstrings...

File: src/compas/scene/volmeshobject.py [hunk 10]
```diff
@@ -72,7 +76,7 @@ def __init__(self, volmesh, **kwargs):
         self.edgewidth = kwargs.get("edgewidth", 1.0)
         self.show_vertices = kwargs.get("show_vertices", False)
         self.show_edges = kwargs.get("show_edges", True)
```
Comment (tomvanmele) [inline src/compas/scene/volmeshobject.py:78]: maybe the default should be False here as well

**Review by tomvanmele — COMMENTED** [review]
should we do this for the blender objects as well?

**Review by tomvanmele — APPROVED** [review]

**Pull request merged**